File: Java/src/_0990_Satisfiability_of_Equality_Equations/Solution.java
package _0990_Satisfiability_of_Equality_Equations;

import java.util.ArrayList;
import java.util.List;

/**
 * Date: 2020/9/17 21:16
 * Content:
 */
public class Solution {

    public boolean equationsPossible(String[] eq) {

        // 第一步：统计字母的个数，并将 eq 根据相等与否分为两个集合
        List<String> equalList = new ArrayList<>();
        List<String> unequalList = new ArrayList<>();
        for (String e : eq) {
            if (e.substring(1, 3).equals("==")) equalList.add(e);
            else unequalList.add(e);
        }
        // 第二步：使用并查集将所有相等的数据放在一起
        UnionFind uf = new UnionFind(26);
        for (String s : equalList)
            uf.unionElements(s.charAt(0) - 'a', s.charAt(3) - 'a');

        // 第三步：根据不相等的集合判断等式方程的可满足性：如果两个数据不相等但却在同一个集合中，则不合法
        for (String s : unequalList) {
            if (uf.isConnected(s.charAt(0) - 'a', s.charAt(3) - 'a'))
                return false;
        }
        return true;
    }

    public static void main(String[] args) {

//        String[] equations = {"c==c", "b==d", "x!=z"};  // true
//        String[] equations = {"a==b", "b!=c", "c==a"};  // false
        String[] equations = {"a==b", "b!=a"};  // false
        System.out.println((new Solution()).equationsPossible(equations));
    }
}

class UnionFind {
    private int[] parent;
    private int[] rank;

    public UnionFind(int size) {
        parent = new int[size];
        rank = new int[size];
        for (int i = 0; i < size; i++) {
            parent[i] = i;
            rank[i] = 1;
        }
    }

    private int find(int p) {
        while (p != parent[p]) {
            parent[p] = parent[parent[p]];
            p = parent[p];
        }
        return p;
    }

    public boolean isConnected(int p, int q) {
        return find(p) == find(q);
    }

    public void unionElements(int p, int q) {
        int pRoot = find(p);
        int qRoot = find(q);

        if (pRoot == qRoot)
            return;

        // 合并
        if (rank[pRoot] < rank[qRoot])
            parent[pRoot] = qRoot;
        else if (rank[qRoot] < rank[pRoot])
            parent[qRoot] = pRoot;
        else {
            parent[pRoot] = qRoot;
            rank[qRoot] += 1;
        }
    }
}

File: Python/_0990_Satisfiability_of_Equality_Equations/Solution.py
# coding=utf-8
# Date: 2021/6/6 16:12
from typing import List


# 执行用时：44 ms, 在所有 Python3 提交中击败了97.54%的用户
# 内存消耗：15 MB, 在所有 Python3 提交中击败了52.46%的用户
class Solution:
    def equationsPossible(self, eq: List[str]) -> bool:
        p = [i for i in range(30)]
        # 第一步：将所有等式对应的字母合并
        for e in eq:
            if e[1:3] == "==":
                a = self.find(p, ord(e[0]) - ord('a'))
                b = self.find(p, ord(e[3]) - ord('a'))
                if a != b:
                    p[a] = b
        # 检查所有的不等式，看是否存在矛盾
        for e in eq:
            if e[1:3] == "!=":
                a = self.find(p, ord(e[0]) - ord('a'))
                b = self.find(p, ord(e[3]) - ord('a'))
                if a == b:
                    return False
        return True

    def find(self, p, x):
        if p[x] != x:
            p[x] = self.find(p, p[x])
        return p[x]


if __name__ == "__main__":
    print(Solution().equationsPossible(["a==b", "b!=a"]))  # False
